File: 095_Snakes and Ladders.cpp
// Snakes and Ladders
// Minimum moves to reach final square on board with snakes and ladders using BFS.
#include <bits/stdc++.h>
using namespace std;

int snakesAndLadders(vector<vector<int>>& board) {
    int n = board.size();
    vector<int> cells(n*n+1);
    int idx=1; int r=n-1,c=0; int dir=1;
    for(int i=n-1;i>=0;--i){ // convert board to 1D
        for(int j=0;j<n;++j){ int col = (dir==1? j: n-1-j); cells[idx++] = board[i][col]; }
        dir *= -1;
    }
    vector<int> dist(n*n+1, -1); queue<int> q; dist[1]=0; q.push(1);
    while(!q.empty()){
        int u=q.front(); q.pop();
        if(u==n*n) return dist[u];
        for(int d=1; d<=6 && u+d<=n*n; ++d){
            int v = u+d; if(cells[v]!=-1) v = cells[v];
            if(dist[v]==-1){ dist[v]=dist[u]+1; q.push(v); }
        }
    }
    return -1;
}

int main(){
    vector<vector<int>> b = {{-1,-1,-1,-1,-1,-1},{-1,-1,-1,-1,-1,-1},{-1,-1,-1,-1,-1,-1},{-1,35,-1,-1,13,-1},{-1,-1,-1,-1,-1,-1},{-1,15,-1,-1,-1,-1}};
    cout<<"Min moves: "<<snakesAndLadders(b)<<"\n";
    return 0;
}

// Time Complexity: O(n^2)
// Space Complexity: O(n^2)

File: 095_Snakes and Ladders.py
"""
Snakes and Ladders
Find minimum moves to win using BFS on 1D-mapped board.
"""
from collections import deque
from typing import List

def snakes_and_ladders(board: List[List[int]]) -> int:
    n = len(board)
    cells = [0]*(n*n+1)
    idx = 1
    for i in range(n-1, -1, -1):
        row = board[i]
        if (n-1-i) % 2 == 0:
            for v in row:
                cells[idx] = v; idx += 1
        else:
            for v in reversed(row):
                cells[idx] = v; idx += 1
    dist = [-1]*(n*n+1)
    dist[1] = 0
    q = deque([1])
    while q:
        u = q.popleft()
        if u == n*n: return dist[u]
        for d in range(1,7):
            v = u + d
            if v > n*n: break
            if cells[v] != -1: v = cells[v]
            if dist[v] == -1:
                dist[v] = dist[u] + 1
                q.append(v)
    return -1

if __name__ == '__main__':
    b = [[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,35,-1,-1,13,-1],[-1,-1,-1,-1,-1,-1],[-1,15,-1,-1,-1,-1]]
    print('Min moves:', snakes_and_ladders(b))

# Time Complexity: O(n^2)
# Space Complexity: O(n^2)
